cmd/kubelet/app/options/globalflags.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package options
+
+import (
+	"flag"
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	// libs that provide registration functions
+	"k8s.io/apiserver/pkg/util/logs"
+	"k8s.io/kubernetes/pkg/version/verflag"
+
+	// ensure libs have a chance to globally register their flags
+	_ "github.com/golang/glog"
+	_ "github.com/google/cadvisor/container/common"
+	_ "github.com/google/cadvisor/container/containerd"
+	_ "github.com/google/cadvisor/container/docker"
+	_ "github.com/google/cadvisor/container/raw"
+	_ "github.com/google/cadvisor/machine"
+	_ "github.com/google/cadvisor/manager"
+	_ "github.com/google/cadvisor/storage"
+)
+
+// AddGlobalFlags explicitly registers flags that libraries (glog, verflag, etc.) register
+// against the global flagsets from "flag" and "github.com/spf13/pflag".
+// We do this in order to prevent unwanted flags from leaking into the Kubelet's flagset.
+func AddGlobalFlags(fs *pflag.FlagSet) {
+	addGlogFlags(fs)
+	addCadvisorFlags(fs)
+	verflag.AddFlags(fs)
+	logs.AddFlags(fs)
+}
+
+// register adds a flag to local that targets the Value associated with the Flag named globalName in global
+func register(global, local *flag.FlagSet, globalName, help string) {
+	if f := global.Lookup(globalName); f != nil {
+		// we should always use hyphens instead of underscores when registering kubelet flags
+		local.Var(f.Value, strings.Replace(globalName, "_", "-", -1), help)
```

Comment:
why not `local.AddFlag(f)`?

Reply:
I hadn't thought of it, actually, though it seems natural now that you mention it.

One justification that comes to mind for `local.Var` instead of `local.AddFlag` is that we get more control of the Kubelet's flags by avoiding any customizations on a third-party flag object, like a `Shorthand`.

But I'll take another look tomorrow when I'm more awake.

Reply:
Turns out https://golang.org/pkg/flag/ doesn't offer an `AddFlag` helper, only pflag offers it. 
I think I still prefer the idea of a clean registration, in either case.

Reply:
Now that I've modified the registration functions to always take a *pflag.FlagSet for local, I can use AddFlag. Avoiding shorthands for converted flags may actually break the API, I realized, because PFlagFromGoFlag already registers a shorthand for one letter flag names. Presumably -v and --v work today. That it also sets NoOptDefVal correctly is a nice bonus. So I think I'll move to AddFlag, since it doesn't seem I have a reason not to at this point.

Reply:
Done